addons/spreadsheet_dashboard/models/spreadsheet_dashboard_share.py

```diff
@@ -0,0 +1,62 @@
+# -*- coding: utf-8 -*-
+import base64
+import uuid
+
+from odoo import models, fields, api, exceptions
+from odoo.tools import consteq
+
+class SpreadsheetDashboardShare(models.Model):
+    _name = 'spreadsheet.dashboard.share'
+    _inherit = 'spreadsheet.mixin'
+    _description = 'Copy of a shared dashboard'
+
+    dashboard_id = fields.Many2one('spreadsheet.dashboard', required=True, ondelete='cascade')
+    excel_export = fields.Binary()
+    access_token = fields.Char(required=True, default=lambda x: str(uuid.uuid4()), groups="base.group_user")
+    full_url = fields.Char(string="URL", compute='_compute_full_url')
+    mode = fields.Char(default="spreadsheet")
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        for vals in vals_list:
+            if "excel_files" in vals:
+                excel_zip = self._zip_xslx_files(
+                    vals["excel_files"]
+                )
+                del vals["excel_files"]
+                vals["excel_export"] = base64.b64encode(excel_zip)
+        return super().create(vals_list)
+
+    @api.depends('access_token')
+    def _compute_full_url(self):
+        for record in self:
+            record.full_url = "%s/dashboard/share/%s/%s?from=%s" % (record.get_base_url(), record.id, record.access_token, record.mode)
+
+    @api.model
+    def action_get_share_url(self, vals):
+        return self.create(vals).full_url
+
+    def _check_token(self, access_token):
+        if not access_token:
+            return False
+        return consteq(access_token, self.access_token)
+
+    def _get_dashboard_and_check_access(self, access_token, operation='read'):
+        self.ensure_one()
+        if not self._check_token(access_token):
+            return False
+        dashboard = self.with_user(self.create_uid).env['spreadsheet.dashboard'].browse(self.dashboard_id.id)
+        if operation == 'write':
+            return self._get_writable_dashboard(dashboard)
+        else:
+            return dashboard
+
+    def _get_writable_dashboard(self, dashboard):
+        self.ensure_one()
+        try:
+            # checks the rights first in case of empty recordset
+            dashboard.with_user(self.env.user).check_access_rights('write')
+            dashboard.with_user(self.env.user).check_access_rule('write')
+        except exceptions.AccessError:
+            return False
+        return dashboard.with_user(self.env.user)._filter_access_rules('write')
```

Comment:
Here are equivalent functions in dashboard that I referred in the task logs. I've already changed them to avoid pass write access right to the internal users who actually don't have.